# HW9

Each of these questions is related to material in Topics 2.2 and 2.3.

In [1]:
name = "Alex"              # Enter your name
user = "XUBK"              # Enter your BC username
print("HW9 submission for {} ({})".format(name,user))

HW9 submission for Alex (XUBK)


Please rename the file so that it includes your BC username. For example, I would change <b>HW9</b> to <b>HW9-reuterj</b> and upload <b>HW9-reuterj.ipynb</b> to Canvas.

In [2]:
import numpy as np

<b>(2 points) 1.</b> Create a (10,) array 'a1' that contains the effective annual interest rates [0.01 0.02 ... 0.09, 0.1 ]. Use one or more steps to create a new (10,) array 'a2' that contains the future value of investing \\$100 for 10 years at each of these rates. 

For example, element 0 of 'a2' should equal 110.46221254 $\left(= \$100(1.01)^{10} \right)$.

In [19]:
arr1 = np.arange(0.01,.11,.01)
arr1

array([0.01, 0.02, 0.03, 0.04, 0.05, 0.06, 0.07, 0.08, 0.09, 0.1 ])

In [21]:
arr2 = 100 * (1+ arr1)**10
arr2

array([110.46221254, 121.899442  , 134.39163793, 148.02442849,
       162.88946268, 179.08476965, 196.71513573, 215.89249973,
       236.73636746, 259.37424601])

<b>(2 points) 2.</b> Use <b>np.random.random()</b> to generate a (3,5) array 'b1' where each element is drawn from uniform (-1,3). Create a new (3,) array 'b2' containing the average values of the three rows. Create a new (5,) array 'b3' containing the average values of the five columns. Count (and print) the number of elements of 'b3' that are positive.

In [102]:
# your code here
b1 = 3 * np.random.random_sample((3,5)) - 1
print("b1 is:")
print(b1)
b2 = np.mean(b1, axis=1)
print("b2 is:")
print(b2)
b3 = np.mean(b1, axis = 0)
print("b3 is:")
print(b3)

b1 is:
[[-0.97131896  1.93291972  0.07631314  1.24110375  1.36313826]
 [-0.88961536 -0.06014803 -0.6623583   0.53807376 -0.28435152]
 [ 0.01696326  1.7104871   1.08680805 -0.58622997  1.15012836]]
b2 is:
[ 0.72843118 -0.27167989  0.67563136]
b3 is:
[-0.61465702  1.1944196   0.16692097  0.39764918  0.7429717 ]


<b>(4 points) 3.</b> 

- Use <b>np.random.binomial()</b> to simulate the outcome of 82 NBA games for a single team. Begin by assuming that the probability of winning each game is 50%. Simulate the 82-game season 10,000 times. Print the 5th, 25th, 50th, 75th, and 95th percentile of the number of games won. 
- Repeat this exercise when the probability of winning equals 45% and then again when it equals 60%. 
- Use the three distributions of number-of-games-won to estimate (i) the probability that a team with a 50% probability of winning wins fewer games in an 82-game seaon than a team with a 45% probability of winning; (ii) the probability that a team a 60% probability of winning wins more games than a team with a 50% probability of winning.

Hint: You can ask how often the number of wins of the 45% team is higher than the number of wins of the 50% team by comparing element 0 of both arrays and then comparing element 1 of both arrays, etc.

In [99]:
# your code here
test = [.5, .45, .6]
p = [5, 25, 50, 75, 95]
dist = []
for x in test:
    s1 = np.random.binomial(82, x, 1000)
    dist.append(s1)
    s1_per = np.percentile(s1, p)
    print("Given {}% win rate - Percentiles: 5th: {} 25th: {} 50th: {} 75th: {} 95th: {}".format(100*x, s1_per[0], s1_per[1], s1_per[2], s1_per[3], s1_per[4]))

Given 50.0% win rate - Percentiles: 5th: 34.0 25th: 38.0 50th: 41.0 75th: 44.0 95th: 48.0
Given 45.0% win rate - Percentiles: 5th: 29.0 25th: 34.0 50th: 37.0 75th: 40.0 95th: 44.0
Given 60.0% win rate - Percentiles: 5th: 42.0 25th: 46.0 50th: 49.0 75th: 52.0 95th: 56.0


In [101]:
num_i = 0
num_ii = 0
for x in range(1000):
    if dist[0][x] < dist[1][x]:
        num_i +=1
    if dist[0][x] < dist[2][x]:
        num_ii +=1
print("Probability that 50% win rate team wins less than a 45% win rate team is {}".format(num_i/1000))
print("Probability that 60% win rate team wins more than a 50% win rate team is {}".format(num_ii/1000))

Probability that 50% win rate team wins less than a 45% win rate team is 0.235
Probability that 60% win rate team wins more than a 50% win rate team is 0.892


<b>(2 points) 4.</b> Use <b>np.random.normal()</b> to generate a (20,) array 'r1' containing draws from the normal distribution, where the mean is 0.10 and the standard deviation is 0.05. Use <b>np.cumprod()</b> to create a new (20,) array 'r2' containing the compound returns implied by 'r1'. The first return in 'r2' corresponds to the 1-year return; the last return in 'r2' corresponds to the 20-year compound return. Print the 20-year compound return.

In [115]:
# your code here
r1 = np.random.normal(0.1, 0.05, 20)
print("r1 is:")
print(r1)
print("r2 is:")
r2 = np.cumprod(1+r1)
print(r2)
print("20 year compound return is: {:.2f}%".format(r2[19]-1))


r1 is:
[0.06000592 0.0863577  0.06243865 0.07300509 0.06119339 0.04328327
 0.15649362 0.06271803 0.07923778 0.11037932 0.07491965 0.09070148
 0.05342968 0.12554534 0.12273656 0.11403112 0.03182741 0.02941342
 0.20104696 0.06653456]
r2 is:
[1.06000592 1.1515456  1.22344655 1.31276438 1.39309688 1.45339467
 1.68084166 1.78626073 1.92780006 2.14058931 2.30096151 2.50966213
 2.64375257 2.9756634  3.3408861  3.7218511  3.84030799 3.95326458
 4.7480564  5.06396625]
20 year compound return is: 4.06%


<b>(1 bonus point).</b> Use 'r2' from question 4 and an additional (20,) array 'r3' of your creation to create new (20,) array 'r4' which contains the geometric average return for years 1 through 20 that are implied by the n-year compound returns in 'r2'. For example, if the first two elements of 'r1' are 0.045 and 0.060, then the 2-year compound return is 0.1077 $(= (1.045)(1.06)-1)$ and the 2-year geometric average return is 0.05247328 $(= ((1.045)(1.06))^{(1/2)} - 1).$ Print the 20-year geometric average return.

In [70]:
# your code here

In [137]:
t1 = np.zeros(2,)
t1[0]= .045
t1[1] = .06
t3 = np.cumprod(1+t1)
t3[1]-1
t2 = t1 +1

In [151]:
(((t3[1])/(t2[1])) * t2[1])**(.5)-1


0.05247327757050435

In [141]:
t2

array([1.045, 1.06 ])

In [167]:
print("20 Year geometric return is: {:2f}%".format(100*((r2[19])**0.5-1)))

20 Year geometric return is: 125.032581%
